src/Nethermind/Nethermind.Core/Extensions/Int64Extensions.cs

```diff
@@ -165,11 +173,32 @@ public static string ToHexString(this in UInt256 value, bool skipLeadingZeros)
                 }
             }
 
-            byte[] bytes = new byte[32];
+            byte[] bytes = GetByteBuffer256();
```

Comment:
I'm guessing, there is some kind of behaviour that prevent the thread from switching to another task in `ToHexString` or something so that the array may not get accidentally reused somewhere else before the `ToHexString` complete?

Reply:
So the `ToHexString` method will complete on the same thread is started on, as no async, callbacks etc.

However since its quite a commonly used path it may be called by multiple threads simultaneously (e.g. multiple different rpc calls being serialised).

To avoid the allocations are several approach that could take:
1. Could rent a buffer from `ArrayPool<byte>.Shared`
2. Could use a `static` single `byte[]` and then `Interlocked.Exchange` with null, so only a single thread will have access to it at any time.
3. Could use a `[ThreadStatic]` `static` which means the static is per thread, so multiple rather than a single copy (though they all start as `null`)

`1.` adds some overhead
`2.` will add contention if there are multiple threads trying to access it at same time, and only save a single allocation if multiple threads are running the same method.
`3.` has no contention and little overheads, however it does cause an allocation per thread in use that calls the method and the object will be long lived.

I choose `3.` as its only a small object (32 bytes) so the long lived allocation is small and the method is likely to be called by multiple threads in parrallel.

The approach is to declare a `static` variable and mark it with the `[ThreadStatic]` attribute:
```cs
[ThreadStatic]
private static byte[]? t_byteBuffer256;
```
That means its now essentially a thread local variable. However if you set it equal to a value in the declaration (when it will be initialised by the static constructor), that will only assign a value on the first call on the first thread and `null` on all other threads.

To get around this then you front it with an accessor method that returns the value **or** if its `null` initialises it and then returns it.
```cs
private static byte[] GetByteBuffer256() => t_byteBuffer256 ??= new byte[32];
```

So now when you call `GetByteBuffer256()` it will allocate the 32 byte buffer once per thread and always return a valid buffer, reusing it between invocations.

However as you rightly point out, you can't use this approach in an async method as the thread may change during the method as it crosses `await`s so would need to use a different approach like 1. or 2.

Reply:
Other way would be changing `ToHexString` to accept `Span<byte>` instead of `byte[]` and just using `stackalloc` here.